src/gdsync.cpp

```diff
@@ -1389,6 +1387,15 @@ struct gds_qp *gds_create_qp(struct ibv_pd *pd, struct ibv_context *context, gds
                 return NULL;
         }
 
+        gds_dbg("before gds_register_peer_ex\n");
+        ret = gds_register_peer_ex(context, gpu_id, &peer, &peer_attr);
+        if (ret) {
+                gds_err("error %d in gds_register_peer_ex\n", ret);
+                goto err_free_cqs;
+        }
+        peer->res_domain = gds_create_res_domain(context);
```

Comment:
can you not move this one back in gds_register_peer_ex()? simply invoke gds_register_peer_ex there unconditionally

Reply:
a preexisting bug, you should also free the res_domain in gds_destroy_qp

Reply:
Because, if gds_create_res_domain() is inside gds_register_peer_ex(), it creates a new res domain not only for the QP but also for the CQs (gds_create_cq() also calls gds_register_peer_ex()). 
I'm not sure if this is ok or there could be some potential issue if for the same peer, every QP and CQ have a different res domain.

Reply:
@drossetti ping

Reply:
if res_domain becomes a property of the QP, then I propose to
- remove res_domain for gds_peer
- add res_domain to gds_cq